crypto/libsodium-fork/src/libsodium/crypto_sign/ed25519/ref10/batch.c

```diff
@@ -0,0 +1,318 @@
+        childl = (parent * 2) + 1;
+        childr = childl + 1;
+    }
+
+    /* sift root back up to its sorted spot */
+    parent = (node - 1) / 2;
+    while (node && lte256_modm_batch(scalars[pheap[parent]], scalars[pheap[node]], limbsize)) {
+        heap_swap(pheap, parent, node);
+        node = parent;
+        parent = (node - 1) / 2;
+    }
+}
+
+/* build the heap with count elements, count must be >= 3 */
+static void
+heap_build(batch_heap *heap, size_t count) {
+    heap->heap[0] = 0;
+    heap->size = 0;
+    while (heap->size < count)
+        heap_insert_next(heap);
+}
+
+/* extend the heap to contain new_count elements */
+static void
+heap_extend(batch_heap *heap, size_t new_count) {
+    while (heap->size < new_count)
+        heap_insert_next(heap);
+}
+
+/* get the top 2 elements of the heap */
+static void
+heap_get_top2(batch_heap *heap, heap_index_t *max1, heap_index_t *max2, size_t limbsize) {
+    heap_index_t h0 = heap->heap[0], h1 = heap->heap[1], h2 = heap->heap[2];
+    if (lt256_modm_batch(heap->scalars[h1], heap->scalars[h2], limbsize))
+        h1 = h2;
+    *max1 = h0;
+    *max2 = h1;
+}
+
+
+/* */
+void ge25519_multi_scalarmult_vartime_final(ge25519_p3 *r, ge25519_p3 *point, sc25519 scalar) { 
+    const sc25519_element_t topbit = ((sc25519_element_t)1 << (SC25519_LIMB_SIZE - 1));
+    size_t limb = limb128bits;
+    sc25519_element_t flag;
+    ge25519_p1p1 p1p1_r;
+    ge25519_cached cache_r;
+
+    if (isone256_modm_batch(scalar)) {
+        /* this will happen most of the time after bos-carter */
+        *r = *point;
+        return;
+    } else if (iszero256_modm_batch(scalar)) {
+        /* this will only happen if all scalars == 0 */
+        memset(r, 0, sizeof(*r));
+        r->Y[0] = 1;
+        r->Z[0] = 1;
+        return;
+    }
+
+    *r = *point;
+
+    /* find the limb where first bit is set */
+    while (!scalar[limb])
+        limb--;
+
+    /* find the first bit */
+    flag = topbit;
+    while ((scalar[limb] & flag) == 0)
+        flag >>= 1;
+
+    
+    ge25519_p3_to_cached(&cache_r, point);
+    /* exponentiate */
+    for (;;) {
+        ge25519_p3_dbl(&p1p1_r, r);
+        ge25519_p1p1_to_p3(r, &p1p1_r);
+        if (scalar[limb] & flag) 
+        {
+            ge25519_add(&p1p1_r, r, &cache_r);
+            ge25519_p1p1_to_p3(r,&p1p1_r);
+        }
+
+
+        flag >>= 1;
+        if (!flag) {
+            if (!limb--)
+                break;
+            flag = topbit;
+        }
+    }
+}
+
+
+/* count must be >= 5 */
+static void
+ge25519_multi_scalarmult_vartime(ge25519_p3 *r, batch_heap *heap, size_t count) {
+    heap_index_t max1, max2;
+
+    /* start with the full limb size */
+    size_t limbsize = SC25519_LIMB_SIZE - 1;
+
+    /* whether the heap has been extended to include the 128 bit scalars */
+    int extended = 0;
+
+    /* grab an odd number of scalars to build the heap, unknown limb sizes */
+    heap_build(heap, ((count + 1) / 2) | 1);
+
+    for (;;) {
```

Comment:
`for (;;)` -> `while(true)`

Reply:
true is not native in C. so we need to define macros for that
if think while(1) is good enough + add some comment.
what do you think?